Using alias directives

When using "Using alias directives" like 
**using ResourcesKeys = xy.Common.Properties.ResourcesKeys;**
in combination with multiple scripts it results in an Exception "error CS1529: A using clause must precede all other elements defined in the namespace except extern alias declarations."
Reason is that it seems that all scripts are merged into one file. All r# and using Statements are cut out except using alias directives. Leading to a script file where an using alias directive is in the middle of two classes.

--- a/src/ScriptCs.ComponentModel.Composition.Test/ScriptCsCatalogFacts.cs
+++ b/src/ScriptCs.ComponentModel.Composition.Test/ScriptCsCatalogFacts.cs
@@ -1,4 +1,4 @@
-﻿using Moq;
+using Moq;
 using ScriptCs.Contracts;
 using Should;
 using System;
@@ -59,6 +59,25 @@ namespace ScriptCs.ComponentModel.Composition.Test
                 var scriptName2 = @"c:\workingdirectory\DoubleScript.csx";
                 var scriptCsCatalog = new ScriptCsCatalog(new[] { scriptName, scriptName2 },
                     GetOptions(fileSystem: GetMockFileSystem(new[] { scriptName, scriptName2 }, new[] { Scripts.SimpleScript, Scripts.DoubleScript }).Object));
+
+                // act
+                var mefHost = GetComposedMefHost(scriptCsCatalog);
+
+                // assert
+                mefHost.Plugins.ShouldNotBeNull();
+                mefHost.Plugins.Count.ShouldEqual(2);
+                mefHost.Plugins[0].DoSomething().ShouldEqual("Simple");
+                mefHost.Plugins[1].DoSomething().ShouldEqual("Double");
+            }
+
+            [Fact]
+            public void ShouldWorkWithMultipleScriptsNotMerged()
+            {
+                // arrange
+                var scriptName = @"c:\workingdirectory\SimpleScript.csx";
+                var scriptName2 = @"c:\workingdirectory\DoubleScript.csx";
+                var scriptCsCatalog = new ScriptCsCatalog(new[] { scriptName, scriptName2 },
+                    GetOptions(fileSystem: GetMockFileSystem(new[] { scriptName, scriptName2 }, new[] { Scripts.SimpleScript, Scripts.DoubleScript }).Object, keepScriptsSeparated: true));
 
                 // act
                 var mefHost = GetComposedMefHost(scriptCsCatalog);
@@ -732,13 +751,14 @@ namespace ScriptCs.ComponentModel.Composition.Test
             return mefHost;
         }
 
-        private static ScriptCsCatalogOptions GetOptions(string[] scriptArgs = null, Type[] references = null, IFileSystem fileSystem = null)
+        private static ScriptCsCatalogOptions GetOptions(string[] scriptArgs = null, Type[] references = null, IFileSystem fileSystem = null, bool keepScriptsSeparated = false)
         {
             return new ScriptCsCatalogOptions
             {
                 FileSystem = fileSystem,
                 References = references,
-                ScriptArgs = scriptArgs
+                ScriptArgs = scriptArgs,
+                KeepScriptsSeparated = keepScriptsSeparated
             };
         }
 

--- a/src/ScriptCs.ComponentModel.Composition.Test/ScriptCsCatalogOptionsFacts.cs
+++ b/src/ScriptCs.ComponentModel.Composition.Test/ScriptCsCatalogOptionsFacts.cs
@@ -57,6 +57,7 @@ namespace ScriptCs.ComponentModel.Composition.Test
                 var options = new ScriptCsCatalogOptions { ScriptArgs = null };
 
                 // act
+
                 var result = options.OverridesNullByDefault();
 
                 // assert
@@ -118,6 +119,20 @@ namespace ScriptCs.ComponentModel.Composition.Test
                 // assert
                 result.FileSystem.ShouldNotBeNull();
                 result.FileSystem.ShouldEqual(fileSystemMock);
+            }
+            
+            
+            [Fact]
+            public void ShouldNotOverridesValuedKeepScriptsSeparated()
+            {
+                // arrange
+                var options = new ScriptCsCatalogOptions { KeepScriptsSeparated = true };
+
+                // act
+                var result = options.OverridesNullByDefault();
+
+                // assert
+                result.KeepScriptsSeparated.ShouldEqual(true);
             }
         }
     }

--- a/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalog.cs
+++ b/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalog.cs
@@ -1,4 +1,4 @@
-﻿using ScriptCs.Contracts;
+using ScriptCs.Contracts;
 using ScriptCs.Hosting;
 using System;
 using System.Collections.Generic;
@@ -404,34 +404,48 @@ namespace ScriptCs.ComponentModel.Composition
 
         private AssemblyCatalog ExecuteScripts(IEnumerable<string> scriptFiles)
         {
-            AssemblyCatalog catalog = null;
-
             var services = CreateScriptServices();
 
-            foreach (var scriptFile in scriptFiles)
+            ScriptResult result = null;
+            if (_options.KeepScriptsSeparated)
             {
-                var loader = GetLoader(scriptFile);
-
-                var result = services.Executor.ExecuteScript(loader);
-
-                if (result.CompileExceptionInfo != null)
+                foreach (var scriptFile in scriptFiles)
                 {
-                    result.CompileExceptionInfo.Throw();
-                }
-
-                if (result.ExecuteExceptionInfo != null)
-                {
-                    result.ExecuteExceptionInfo.Throw();
-                }
-
-                var marker = result.ReturnValue as Type;
-
-                if (marker != null && catalog == null)
-                {
-                    catalog = new AssemblyCatalog(marker.Assembly, this);
+                    var loader = GetLoader(new[] { scriptFile });
+                    result = ExecuteScript(services, loader);
                 }
             }
+            else
+            {
+                var loader = GetLoader(scriptFiles);
+                result = ExecuteScript(services, loader);
+            }
+
+            AssemblyCatalog catalog = null;
+            var marker = result.ReturnValue as Type;
+            if (marker != null)
+            {
+                catalog = new AssemblyCatalog(marker.Assembly, this);
+            }
+
             return catalog;
+        }
+
+        private ScriptResult ExecuteScript(ScriptServices services, string loader)
+        {
+            var result = services.Executor.ExecuteScript(loader);
+
+            if (result.CompileExceptionInfo != null)
+            {
+                result.CompileExceptionInfo.Throw();
+            }
+
+            if (result.ExecuteExceptionInfo != null)
+            {
+                result.ExecuteExceptionInfo.Throw();
+            }
+
+            return result;
         }
 
         private string GetFullPath(string path)
@@ -459,11 +473,14 @@ namespace ScriptCs.ComponentModel.Composition
             return string.Format(CultureInfo.CurrentCulture, "{0} (Path=\"{1}\")", GetType().Name, Path);
         }
 
-        private string GetLoader(string scriptFile)
+        private string GetLoader(IEnumerable<string> scriptFiles)
         {
             var builder = new StringBuilder();
 
-            builder.AppendFormat("#load {0}{1}", scriptFile, Environment.NewLine);
+            foreach (var scriptFile in scriptFiles)
+            {
+                builder.AppendFormat("#load {0}{1}", scriptFile, Environment.NewLine);
+            }
             builder.AppendLine("public class Marker {}");
             builder.AppendLine("typeof(Marker)");
 

--- a/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalog.cs
+++ b/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalog.cs
@@ -404,30 +404,33 @@ namespace ScriptCs.ComponentModel.Composition
 
         private AssemblyCatalog ExecuteScripts(IEnumerable<string> scriptFiles)
         {
+            AssemblyCatalog catalog = null;
+
             var services = CreateScriptServices();
 
-            var loader = GetLoader(scriptFiles);
-
-            var result = services.Executor.ExecuteScript(loader);
-
-            if (result.CompileExceptionInfo != null)
+            foreach (var scriptFile in scriptFiles)
             {
-                result.CompileExceptionInfo.Throw();
+                var loader = GetLoader(scriptFile);
+
+                var result = services.Executor.ExecuteScript(loader);
+
+                if (result.CompileExceptionInfo != null)
+                {
+                    result.CompileExceptionInfo.Throw();
+                }
+
+                if (result.ExecuteExceptionInfo != null)
+                {
+                    result.ExecuteExceptionInfo.Throw();
+                }
+
+                var marker = result.ReturnValue as Type;
+
+                if (marker != null && catalog == null)
+                {
+                    catalog = new AssemblyCatalog(marker.Assembly, this);
+                }
             }
-
-            if (result.ExecuteExceptionInfo != null)
-            {
-                result.ExecuteExceptionInfo.Throw();
-            }
-
-            var marker = result.ReturnValue as Type;
-
-            AssemblyCatalog catalog = null;
-            if (marker != null)
-            {
-                catalog = new AssemblyCatalog(marker.Assembly, this);
-            }
-
             return catalog;
         }
 
@@ -456,14 +459,11 @@ namespace ScriptCs.ComponentModel.Composition
             return string.Format(CultureInfo.CurrentCulture, "{0} (Path=\"{1}\")", GetType().Name, Path);
         }
 
-        private string GetLoader(IEnumerable<string> scriptFiles)
+        private string GetLoader(string scriptFile)
         {
             var builder = new StringBuilder();
-            foreach (var script in scriptFiles)
-            {
-                builder.AppendFormat("#load {0}{1}", script, Environment.NewLine);
-            }
 
+            builder.AppendFormat("#load {0}{1}", scriptFile, Environment.NewLine);
             builder.AppendLine("public class Marker {}");
             builder.AppendLine("typeof(Marker)");
 

--- a/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalogOptions.cs
+++ b/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalogOptions.cs
@@ -40,7 +40,8 @@ namespace ScriptCs.ComponentModel.Composition
                 References = References,
                 ScriptArgs = ScriptArgs ?? new string[0],
                 FileSystem = FileSystem ?? new FileSystem(),
-                Modules = Modules ?? new string[0]
+                Modules = Modules ?? new string[0],
+                KeepScriptsSeparated = KeepScriptsSeparated
             };
         }
     }

--- a/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalogOptions.cs
+++ b/src/ScriptCs.ComponentModel.Composition/ScriptCsCatalogOptions.cs
@@ -28,6 +28,11 @@ namespace ScriptCs.ComponentModel.Composition
         /// </summary>
         public string[] Modules { get; set; }
 
+        /// <summary>
+        /// Prevents the scripts from beeing merged.
+        /// </summary>
+        public bool KeepScriptsSeparated { get; set; }
+
         internal ScriptCsCatalogOptions OverridesNullByDefault()
         {
             return new ScriptCsCatalogOptions

